parrot.rs

```diff
@@ -0,0 +1,73 @@
+use core::sync::atomic::{AtomicUsize, Ordering};
+use kernel::{
+    bindings::msleep,
+    chrdev::Registration,
+    file::{File, Operations, ToUse},
+    io_buffer::IoBufferWriter,
+    prelude::*,
+};
+
+mod frames;
+use frames::*;
+
+module! {
+    type: ParrotSafe,
+    name: b"party_parrot",
+    author: b"John Baublitz (john.m.baublitz@gmail.com)",
+    description: b"Get the party started",
+    license: b"GPL",
+}
+
+const FRAMES: [&str; 10] = [
+    FRAME0, FRAME1, FRAME2, FRAME3, FRAME4, FRAME5, FRAME6, FRAME7, FRAME8, FRAME9,
+];
+
+struct ParrotSafe(Pin<Box<Registration<1>>>);
+
+impl kernel::Module for ParrotSafe {
+    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
+        let mut registration = Registration::new_pinned(name, 0, module)?;
+        registration.as_mut().register::<ParrotOps>()?;
+        Ok(ParrotSafe(registration))
+    }
+}
+
+struct ParrotOps(usize);
+
+impl<'a> Operations for ParrotOps {
+    const TO_USE: ToUse = ToUse {
```

Comment:
You can use:

```rust
kernel::declare_file_operations!(read);
```

Instead of the verbose definition you have now.

Once https://github.com/Rust-for-Linux/linux/pull/322 is merged, you'll be able to replace it with a `#[vtable]` attribute.